File: solutions/elixir/012-longest-common-prefix.ex
# Problem 12: Longest Common Prefix (LeetCode 14)
# Difficulty: Easy
# Language: Elixir
# 
defmodule LongestCommonPrefix do
  def longest_common_prefix([]), do: ""
  def longest_common_prefix(strs) do
    strs
    |> Enum.reduce(fn str, prefix ->
      prefix
      |> String.graphemes()
      |> Enum.zip(String.graphemes(str))
      |> Enum.take_while(fn {a, b} -> a == b end)
      |> Enum.map(&elem(&1, 0))
      |> Enum.join()
    end)
  end
end

File: solutions/typescript/012-longest-common-prefix.ts
/**
 * Problem 12: Longest Common Prefix (LeetCode 14)
 * Difficulty: Easy
 * Language: TypeScript
 */
function longestCommonPrefix(strs: string[]): string {
  if (strs.length === 0) return '';
  return strs.reduce((prefix, str) =>
    [...prefix].reduce(
      (acc, ch, i) => (i < str.length && str[i] === ch ? acc + ch : acc),
      ''
    ).slice(0, [...prefix].findIndex((ch, i) =>
      i >= str.length || str[i] !== ch
    ) === -1 ? prefix.length : [...prefix].findIndex((ch, i) =>
      i >= str.length || str[i] !== ch
    ))
  );
}
